### PR TITLE
SOLR-16203: Initialize schema plugins loaded by SPI name

### PR DESCRIPTION
**Pull request by janhoy**
https://issues.apache.org/jira/browse/SOLR-16203

### CODE REVIEW

**Comment by janhoy** [comment]
First attempt in efadf94dc6ac87027ef708c71c58d3848e7a9a0d:

Naïve solution, explicitly calling `inform()` in `FieldTypePluginLoader`. The two tests from JIRA are added to this PR to demonstrate. 

There may be a more elegant place to do the init. We risk double inform if not using SPI since we don't check that here. Perhaps there is some method somewhere that could be called on the entire chain instead of in each init(), such as the one in `ManagedIndexSchema#informResourceLoaderAwareObjectsInChain()`. I did not dive deep here. Open to suggestions from folks who know the code better.

**Comment by janhoy** [comment]
Second attempt:

The first attempt did not work for packages, since `FieldTypePluginLoader#init` was too early in the cycle to initialize analyzers from packages.

Managed schema handled the missing `inform` call after SPI already, so moved the method `informResourceLoaderAwareObjectsForFieldType` from `ManagedIndexSchema` to `IndexSchema` and use it at the end of `IndexSchema#readSchema`. All tests pass.

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (uschindler) [inline solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2001]: does this need to be protected? The inner method is private, so why is this one public? A method reference in lambdas does not need to be visible in any way.

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (uschindler) [inline solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: maybe do something like:

```java
Stream.of(fieldType.getIndexAnalyzer(), fieldType.getQueryAnalyzer(), noIdeaHowToHandleThis.getMultiTermAnalyzer()).filter(a -> a instanceof TokenizerChain)
  .distinct().foreach(a -> informResourceLoaderAwareObjectsInChain((TokenizerChain) a);
```

Comment (uschindler) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: maybe this one works:

```java
var stream = Stream.of(fieldType.getIndexAnalyzer(), fieldType.getQueryAnalyzer());
if (fieldType instanceof TextField) {
  stream = Stream.concat(stream, Stream.of(((TextField) fieldType).getMultiTermAnalyzer());
}
stream.filter(a -> a instanceof TokenizerChain)
  .distinct().foreach(a -> informResourceLoaderAwareObjectsInChain((TokenizerChain) a);
```

Comment (uschindler) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: (the distinct is by equals, but by default they use system's equals)

Comment (uschindler) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: null check is not needed because of `TokenizerChain.class::isInstance` (returns false on `null`)

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2001]: This is still also called from `ManagedIndexSchema` class which is a subclass.

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: Note that this entire method is *moved* over from `ManagedIndexSchema`, and I have not modified it or attempted to optimize it. May we leave improvements to anoter PR?

Comment (uschindler) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: Ah, you're right. I just thought al little bit of deduplication of code would be fine.

It's up to you to apply that, but I can make a separate PR.

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (uschindler) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2001]: This is not ideal. If you move the call to this private method into `postReadInfom()` of *this* class (`IndexSchema`), then the subclass would also call it and it could be private.
Here it is called in readSchema *after* `postReadInform()`. If it is moved there both implementations automatically use it.

**Review by uschindler — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (uschindler) [inline solr/core/src/java/org/apache/solr/schema/IndexSchema.java:660]: move this line into the `postReadInform()` method called before.

File: solr/core/src/java/org/apache/solr/schema/ManagedIndexSchema.java
Comment (uschindler) [inline solr/core/src/java/org/apache/solr/schema/ManagedIndexSchema.java:1299]: remove this whole method, too

**Review by uschindler — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/schema/IndexSchema.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/schema/IndexSchema.java:2006]: I'll close this comment and defer to new PR after this landed

**Comment by janhoy** [comment]
Thanks for nice proposals @uschindler , the method is now private and code cleaner

**Review by uschindler — APPROVED** [review]

**Pull request merged**
